model.py

```diff
@@ -1,108 +1,144 @@
-            child_h = Var(torch.zeros(1, 1, self.mem_dim))
-            if self.cudaFlag:
-                child_c, child_h = child_c.cuda(), child_h.cuda()
+class Tree(object):
+    def __init__(self, idx):
+        self.children = []
+        self.idx = idx
+
+    def __repr__(self):
+        if self.children:
+            return '{0}: {1}'.format(self.idx, str(self.children))
+        else:
+            return str(self.idx)
+
+tree = Tree(0)
+tree.children.append(Tree(1))
+tree.children.append(Tree(2))
+tree.children.append(Tree(3))
+tree.children[1].children.append(Tree(4))
+print(tree)
+
+class ChildSumLSTMCell(nn.Module):
+    def __init__(self, hidden_size,
+                 i2h_weight_initializer=None,
+                 hs2h_weight_initializer=None,
+                 hc2h_weight_initializer=None,
+                 i2h_bias_initializer='zeros',
+                 hs2h_bias_initializer='zeros',
+                 hc2h_bias_initializer='zeros',
+                 input_size=0):
+        super(ChildSumLSTMCell, self).__init__()
+        self._hidden_size = hidden_size
+        self._input_size = input_size
+        stdv = 1. / math.sqrt(input_size)
+        self.i2h_weight = Parameter(torch.Tensor(4*hidden_size, input_size).uniform_(-stdv, stdv))
+        self.i2h_bias = Parameter(torch.Tensor(4*hidden_size).uniform_(-stdv, stdv))
+        stdv = 1. / math.sqrt(hidden_size)
+        self.hs2h_weight = Parameter(torch.Tensor(3*hidden_size, hidden_size).uniform_(-stdv, stdv))
+        self.hs2h_bias = Parameter(torch.Tensor(3*hidden_size).uniform_(-stdv, stdv))
+        stdv = 1. / math.sqrt(hidden_size)
+        self.hc2h_weight = Parameter(torch.randn(hidden_size, hidden_size).uniform_(-stdv, stdv))
+        self.hc2h_bias = Parameter(torch.Tensor(hidden_size).uniform_(-stdv, stdv))
+
+    def forward(self, inputs, tree):
+        children_outputs = [self(inputs, child) for child in tree.children]
+        if children_outputs:
+            _, children_states = zip(*children_outputs) # unzip
+        else:
+            children_states = None
+
+        return self.node_forward(inputs[tree.idx].unsqueeze(0),
+                                 children_states,
+                                 self.i2h_weight, self.hs2h_weight,
+                                 self.hc2h_weight, self.i2h_bias,
+                                 self.hs2h_bias, self.hc2h_bias)
+
+    def node_forward(self, inputs, children_states,
+                     i2h_weight, hs2h_weight, hc2h_weight,
+                     i2h_bias, hs2h_bias, hc2h_bias):
+        # comment notation:
+        # N for batch size
+        # C for hidden state dimensions
+        # K for number of children.
+
+        # FC for i, f, u, o gates (N, 4*C), from input to hidden
+        i2h = F.linear(inputs, i2h_weight, i2h_bias)
+        i2h_slices = torch.split(i2h, i2h.size(1) // 4, dim=1) # (N, C)*4
+        i2h_iuo = torch.cat([i2h_slices[0], i2h_slices[2], i2h_slices[3]], dim=1) # (N, C*3)
```

Comment:
Why are the indices 0,2,3 and not 0,1,2? Why is `i2h_f_slice = i2h_slices[1]`?

Reply:
Also why is it `iuo` and not `iou`? Is there some rationale behind this? I am using `iou` so wondering if I am making some mistake...